### PR TITLE
[tf.embedding_column] support tf.sparse.segment_mean

### PR DESCRIPTION
**Pull request by Orion34-lanbo**

### CODE REVIEW

File: tao_compiler/mlir/disc/IR/hlo_disc_ops.td [hunk 1]
```diff
@@ -61,6 +61,10 @@ def HLO_TensorOrIndexTensor : AnyTypeOf<[HLO_Tensor, IndexTensor]>;
 
 def HLO_I32Tensor : TensorOf<[SignlessIntOfWidths<[32]>]>;
 
+def HLO_I64Tensor : TensorOf<[SignlessIntOfWidths<[64]>]>;
+
+def HLO_IndexTensor : AnyTypeOf<[HLO_I32Tensor, HLO_I64Tensor]>;
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/IR/hlo_disc_ops.td:66]: why `IndexTensor` is not `Index` type but `I32` and `I64`?

File: tao_compiler/mlir/disc/IR/hlo_disc_ops.td [hunk 2]
```diff
@@ -238,4 +242,77 @@ def HLO_SparseFillEmptyRowsOp : HLODISC_ShapedInterfaceOp<"sparse_fill_empty_row
   );
 }
 
+// op {
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/IR/hlo_disc_ops.td:245]: Please remove useless comments.

File: tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc [hunk 3]
```diff
@@ -4098,6 +4098,292 @@ LogicalResult lowerWithScheduleSparseFillEmptyRowsOpCPU(
   return success();
 }
 
+LogicalResult lowerWithScheduleSparseSegmentMeanOpCPU(
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc:4101]: nit: Could all sparse segment reduction op reuse skeleton of this function?

File: tao_compiler/mlir/disc/IR/hlo_disc_ops.td (hunk 1 above)
Comment (wyzero) [inline tao_compiler/mlir/disc/IR/hlo_disc_ops.td:66]: `index` type has special meaning in MLIR, maybe use `I32Or64Tensor`?

File: tao_compiler/mlir/disc/tests/tensorflow_ops/sparse_segment_mean.cc [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      /*num_outputs*/ 1,
+      /*input_Xescriptors*/ {"4xf32_X", "6xi32_X", "6xi32_X"},
+      /*output_Xescriptors*/ {"f32_X"},
+      /*input_vals*/ {{5.0, 6, 7, 8}, {0, 1, 2, 0, 1, 3}, {0, 0, 1, 2, 2, 2}}));
+}
+
+TEST(TFSparseSegmentMeanOpTest, StaticShapeF32Dim1Input) {
+  EXPECT_TRUE(feature_test_main(
+      /*mlir_file_path*/ c_ft_path + "sparse_segment_mean_s_f32_1d.mlir",
+      /*backend_types*/ {BackendType::kX86},
```
Comment (wyzero) [inline tao_compiler/mlir/disc/tests/tensorflow_ops/sparse_segment_mean.cc:50]: better to add AArch64 as well.

File: tao_compiler/mlir/disc/IR/hlo_disc_ops.td (hunk 1 above)
Comment (Orion34-lanbo) [reply tao_compiler/mlir/disc/IR/hlo_disc_ops.td]: fixed

File: tao_compiler/mlir/disc/tests/tensorflow_ops/sparse_segment_mean.cc (hunk 4 above)
Comment (Orion34-lanbo) [reply tao_compiler/mlir/disc/tests/tensorflow_ops/sparse_segment_mean.cc]: done

File: tao_compiler/mlir/disc/IR/hlo_disc_ops.td (hunk 1 above)
Comment (Orion34-lanbo) [reply tao_compiler/mlir/disc/IR/hlo_disc_ops.td]: change to `I32OrI64Tensor`

File: tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc (hunk 3 above)
Comment (Orion34-lanbo) [reply tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc:4101]: It is possible, only the reduction part is different

File: tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc [hunk 5]
```diff
@@ -4098,6 +4098,248 @@ LogicalResult lowerWithScheduleSparseFillEmptyRowsOpCPU(
...
+    segment_count_memref = alloc.getResult();
+
+    {
+      auto for_op =
+          b.create<scf::ForOp>(loc, /* lowerBound */ zero,
+                               /* upperBound */ num_results, /* step */ one);
+      for_op.getBody()->clear();
+      b.setInsertionPointToStart(for_op.getBody());
+      Value i = for_op.getInductionVar();
+      b.create<memref::StoreOp>(loc, zero_floating, segment_count_memref, i);
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc:4164]: Just curious could this for loop be translated to a true `memset` after lowering ?

**Review by qiuxiafei — APPROVED** [review]
LGTM

File: tao_compiler/mlir/disc/transforms/tests/cpu-only-sparse-segment-mean-lhlo-legalize-roots-to-loops.mlir [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+// CHECK-SAME: (%[[INPUT1:.*]]: memref<?x?xf32, "cpu">, %[[INPUT2:.*]]: memref<?xi32, "cpu">, %[[INPUT3:.*]]: memref<?xi32, "cpu">, %[[OUT1:.*]]: memref<?x?xf32, "cpu">) -> memref<?x?xf32, "cpu">
+func.func @sparse_segment_mean(
+    %input1: memref<?x?xf32, "cpu">,
+    %input2: memref<?xi32, "cpu">,
+    %input3: memref<?xi32, "cpu">,
+    %out1: memref<?x?xf32, "cpu">
+    ) -> (
+    memref<?x?xf32, "cpu">
+  ) {
+  // CHECK-NOT lmhlo
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/tests/cpu-only-sparse-segment-mean-lhlo-legalize-roots-to-loops.mlir:13]: Is the following check strong enough?

Comment (Orion34-lanbo) [reply tao_compiler/mlir/disc/transforms/tests/cpu-only-sparse-segment-mean-lhlo-legalize-roots-to-loops.mlir:13]: Current impl has 2 major loops(the one doing memset not included), this checks for the most complex one which dose the naive mean computation.

**Pull request merged**